### DISCUSSION TITLE
Network layout engine

### DISCUSSION THREAD

**Discussion by mihai-sysbio** [MetabolicAtlas/MetabolicAtlas#539] [Ideas]
#### Introduction

Since genome scale metabolic models include many thousands of reactions, metabolites and genes, one often has to select a small part of these to focus on. Visually, this is achieved on Metabolic Atlas with the help of the 2D and 3D viewers. The main difference between these viewers is, surprisingly, not the extra dimension - it is the process through which they were obtained.

The layout of the 2D maps are obtained by manually arranging the nodes of the network so that they are visually appealing. This is a very intensive process with regards to workload. Another downside is that the layout cannot be reused easily, since each layout is obtained so as to maximize the visual appeal. That being said, some tools have added [functionality to reuse maps](https://escher.readthedocs.io/en/latest/tips-and-tricks.html#escher-performance-with-large-maps). Nevertheless, the research community has been using 2D maps for a long time, as these maps are very well suited for printed works. In the latest release, Metabolic Atlas presents users with about 170 maps of various sizes.

The layout of the 3D maps, on the other hand, is the result of a force field algorithm, with parameters tweaked for the size of the network and visual appeal. The process of obtaining a layout is measured in seconds. However, the 3D maps are much harder to use in static formats, as images.

There exist algorithms for automatically creating 2D network layouts. These tend to be embedded in different tools, such as Cytoscape. However, the visual appeal of the resulting layouts is quickly lost with the increase of the size of the network. Moreover, it is often that the layouts do not make the best use of space.

On Metabolic Atlas, the maps are grouped by category: compartment, subsystem and custom. The compartment maps can be very large. For example, the cytosol maps for Human-GEM had to be split into 5 parts. The subsystem and custom maps, on the other hand, can be quite small. The difference is size indicates that the network layout engine can have different strategies based on the size of the map.

This idea is focused on creating an engine that, given a network, it returns a _beautiful_ layout without manual intervention.

#### References

Recent layout engines:
- [CartoGRAPHs](https://github.com/menchelab/CartoGRAPHs)

Here are some references for editors and layout engines:
- [Adaptagrams](http://www.adaptagrams.org/) includes [automatic diagram layout](https://github.com/mjwybrow/adaptagrams/)
- [COPASI](http://copasi.org/Support/Features/)
- [PyRKViewer](https://github.com/evilnose/PyRKViewer)
- [Graphviz](https://graphviz.org/)
- Figure 1 in [this research article](https://journals.plos.org/plosone/article?id=10.1371/journal.pone.0007710)
- [Chisio](http://www.cs.bilkent.edu.tr/~ivis/chisio.html)
- [MetaboMAPS](https://f1000research.com/articles/9-288/v2)
- [iPath3.0](https://academic.oup.com/nar/article/46/W1/W510/4990021)
- [Newt](http://newteditor.org), [source code](https://github.com/iVis-at-Bilkent/newt)
- [Cytoscape.js](https://js.cytoscape.org/) and plugins such as [BiNoM](https://binom.curie.fr/)
- [yEd](https://www.yworks.com/products/yed) with [ySBGN](https://github.com/sbgn/ySBGN)
- [MINERVA](https://minerva.pages.uni.lu/doc/)
- [NaviCell](https://navicell.vincent-noel.fr)
- [hiPathia](http://hipathia.babelomics.org/)
- [Metabolizer](http://metabolizer.babelomics.org/)
- [eQuilibrator](https://equilibrator.weizmann.ac.il/)
- [Escher](https://escher.github.io/), [source code](https://github.com/zakandrewking/escher)
- [MetExploreViz](https://metexplore.toulouse.inrae.fr/metexploreViz/doc/index.php)
- [Caffeine](https://caffeine.dd-decaf.eu/)
- [Pathway Tools](https://bioinformatics.ai.sri.com/ptools/) part of [BioCyc](https://biocyc.org/PToolsWebsiteHowto.shtml#route_search)
- [Omix Visualization](http://www.omix-visualization.com/)
- [NetworkAnalyst](https://www.networkanalyst.ca/)
- [SBNE](https://sbne.readthedocs.io/en/latest/overview.html) and [GH repo](https://github.com/adelhpour/SBNE)
- [Model Explorer](https://github.com/TheAngryFox/ModelExplorer) has a table with the FPS comparison between models. Their tool is implemented in C for desktop use.
- [Escher.jl](https://github.com/stelmo/Escher.jl)
- [SyBLaRS](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1010635)

And some more references for maps, mainly 2D:
- [BRENDA maps](https://www.brenda-enzymes.org/pathway_index.php)
- [KEGG Atlas](https://pubmed.ncbi.nlm.nih.gov/18477636/)
- [Disease maps](http://disease-maps.org)
- [NDEx](http://www.ndexbio.org/#/)
- [SBGN maps](https://sbgn.github.io/publishedmaps) and [sbgnShiny](https://github.com/cannin/sbgnvizShiny)
- [WikiPathways](https://www.wikipathways.org/index.php/WikiPathways)
  - [PathVisio](https://pathvisio.org)
- [KnetMaps.js](https://github.com/Rothamsted/knetmaps.js)

References for formats:
- [SBGN](https://sbgn.github.io/learning), [2](https://kim25.wwwdns.kim.uni-konstanz.de/vanted/addons/sbgn-ed/), [3](https://sbgn.github.io/referencecards), [4](https://github.com/sbgn/sbgn/wiki)
- [KGML](https://www.kegg.jp/kegg/xml)
- WikiPathway [GPML](https://pathvisio.org/documentation/GPML.html)
- as supported by [Cytoscape](https://manual.cytoscape.org/en/stable/Supported_Network_File_Formats.html)

#### Open questions

A proposal should address the following concerns:
- What information is needed to generate a beautiful layout?
- What is a good way to store the obtained layout?
- What if the engine occasionally returns ugly layouts?
- What would make the engine easy to integrate with the existing Metabolic Atlas code base?
- What computational resources are needed to obtain the layout?
- How can the file format be easy to reuse (integrate with other tools)?

**Comment by mihai-sysbio** [comment]
@e0 mentioned this potentially useful tool https://ivanceras.github.io/svgbob-editor.

**Comment by nanjiangshu** [comment]
Well done @mihai-sysbio for excellent description of the new engine for 2D maps. 

If I understand correctly, what we want is a tool (or method, or pipeline or program) that uses the metabolism network as an input and generate a 2D map usable by 2D Map viewer of MetAtlas automatically, that is, without manual intervention.

If that is true, then, the sentence `This idea is focused on creating an engine that, given a network, it returns a beautiful layout.` in the introduction is not quite precise, since this statement does not exclude manual intervention when generating the 2D maps. 

If I remember correctly, the procedure for which current 2D maps were obtained has been described somewhere. It will be nice to include a link in this document so that one can easily refer to.

**Reply by mihai-sysbio** [reply]
Thanks for the valuable feedback. I've amended the line to include `without manual intervention`.
The workflow has been described in wiki of a private repository in SysBio: https://github.com/SysBioChalmers/OMIX_plugin/wiki/Workflow . While we should improve on that work, we shouldn't see it as a guide that needs automation - instead, we should take the opportunity to think from the ground up.

**Comment by johaGL** [comment]
Hello everyone here, I am a bioinformatician and we are also interested in obtaining maps with aesthetics that are satisfactory for publication.  Many of the things you have posted here have been also part of our discussions.  For now or repos are not public as they are still in the very beggining. We have preferred Human Metabolic Atlas as database for this project (we cite you everywhere,  thank you for your great job). Our teams do not know each other, but we will be happy for any collaborative activity, if you are interested too. Best regards. Johanna

**Reply by e0** [reply]
Hi Johanna, thanks for taking time to post here! It would be really fun to follow the progress of your project. If you want any extra input during the development phase, feel free to invite me as a collaborator. I can't promise any active development but I'm always happy to discuss and provide input if that would help, especially about web development details.

I should add that I'm not officially associated with Metabolic Atlas at the moment although I have been an active contributor over the past few years. Anyhow it's still an open source project that I care a lot about. You probably will hear soon from the Metabolic Atlas team too.

**Reply by johaGL** [reply]
Hello @e0 , thank you, I'm very happy to read all this!. I will be back to you/Metabolic Atlas team later. Kind regards :).

**Reply by mihai-sysbio** [reply]
Hi Johanna! Very nice to hear about your ongoing work, it would be great to sync efforts. Please feel free to email me directly at mihail.anton@chalmers.se.
